streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -347,7 +355,11 @@ public void print(final Printed<K, V> printed) {
 
         mergeNode.setMergeNode(true);
         builder.addGraphNode(Arrays.asList(this.streamsGraphNode, streamImpl.streamsGraphNode), mergeNode);
-        return new KStreamImpl<>(builder, name, allSourceNodes, requireRepartitioning, mergeNode);
+
+        // inherit from both streams, with this stream taking higher priority
```

Comment:
I've thought about that, but it seems not possible. Will try out and see if I can do sth. more.

Reply:
I've tried sth. but java.lang.Class does not really have sufficient interfaces. @vvcephei do you have any ideas?

Reply:
As my comment said

> Not sure if we can compare in a useful way to be honest due to type erasure?

Maybe we can't do better as-is. It just seems a little arbitrary to pick the first one if there are two Serdes -- not might still be the best solution though. Ie, I am not happy with the code but don't have a suggestion how we could improve it either.

Reply:
This is an interesting question.

One low-fi solution would be to think about using `equals()` (I think to pull this off, we'd need to introduce a requirement that serde/serializer/deserializers implement equals in a way that would be semantically sound for us. This would not be a back-ward compatible change.

On the other hand, since callers actually subclass `Serde<T>` with a fixed type like `Serde<String>`, it actually should be available at runtime. I don't remember the hoops you have to jump through to get it right now, but I'll revisit it tomorrow.

Reply:
Ok, I've done enough kicking the tires to say that it is possible to get the generic type argument to Serde in most (but not all) cases. It's not a trivial algorithm. I'd estimate it at one day of work. If the current code is sufficient, then I recommend we go with it in this PR and have a follow-up PR for just this feature.

While researching the issue, I found this article that pretty accurately describes what's going on: https://www.javacodegeeks.com/2013/12/advanced-java-generics-retreiving-generic-type-arguments.html

Reply:
Actually, I think we want to err on the side of correctness, so shouldn't we just drop the serdes at this point and set them to null? It seems like this would preserve the current behavior (where we also don't have serdes for the resulting k/v).

Reply:
@guozhangwang The code was not update. Is this intentional? What about John's suggestion to be conservative and drop both serdes for now and fix up later?

Reply:
My bad, it's meant to be dropped.